### ISSUE TITLE
[BUG]  When using Speedup V2, there will be a problem of memory leakage

### ISSUE THREAD

**Issue by lianxintao** [microsoft/nni#5482]
**Describe the bug**:
 When I use NNI's AMC PRUNER to prune YOLOV5, the self.prpagate_originally (), self.Update_direct_sparsity (), self.Update_Sparsity () of the speedup will occupy a lot memory.  And when the Speedup ends, these memory will not be completely released. Because of this problem, after running Speedup multiple times, memory overflow will occur

**Environment**:       YOLOV5 + COCO dataset + speedup v2
- NNI version:   latest  of  the master branch 
- Python version: python3.9
- PyTorch version:   1.10
- Cpu or cuda version: cuda

**Reproduce the problem**
- Code|Example:

- How to reproduce:

**Comment by J-shang** [comment]
hello @lianxintao, does `torch.cuda.empty_cache()` work?

**Comment by lianxintao** [comment]
no，it doesn’t work. I also tried removing the self.node_info in SpeedUp or creating a new process to speedup, but neither of them solved the problem.

**Comment by lianxintao** [comment]
here is a  log generte by memory_profiler

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
   411 6032.9180 MiB 6032.9180 MiB           1       @profile(precision=4, stream=open("/lxt_projects/memory_profiler.log", "w+"))
   412                                             def speedup_model(self) -> GraphModule:
   413 6032.9180 MiB   0.0000 MiB           1           try:
   414 6032.9180 MiB   0.0000 MiB           1               ori_state_dict_file = tempfile.NamedTemporaryFile(delete=False)
   415 6033.0703 MiB   0.1523 MiB           1               torch.save(self.graph_module.state_dict(), ori_state_dict_file)
   416 6033.0703 MiB   0.0000 MiB           1               ori_state_dict_file.close()
   417                                         
   418 6033.0703 MiB   0.0000 MiB           1               self.logger.info("Start to speedup the model...")
   419 6033.0703 MiB   0.0000 MiB           1               training = self.graph_module.training
   420 6033.0703 MiB   0.0000 MiB           1               self.graph_module.train(False)
   421                                         
   422                                                     # TODO: suppose to fix the conflict after the sparsity propagation, which is more elegent
   423 6033.0703 MiB   0.0000 MiB           1               self.logger.info('Resolve the mask conflict before mask propagate...')
   424 6299.4141 MiB 266.3438 MiB           1               fix_mask_conflict(self.masks, self.graph_module, self.dummy_input)
   425 6299.4141 MiB   0.0000 MiB           1               self.logger.info('Infer module masks...')
   426 6299.4141 MiB   0.0000 MiB           1               self.initialize_propagate(self.dummy_input)
   427 16667.4258 MiB 10368.0117 MiB           1               self.propagate_originally()
   428 16667.4258 MiB   0.0000 MiB           1               self.initialize_update_sparsity()
   429 19444.4570 MiB 2777.0312 MiB           1               self.update_direct_sparsity()
   430 23269.9570 MiB 3825.5000 MiB           1               self.update_indirect_sparsity()
   431 23269.9570 MiB   0.0000 MiB           1               self.logger.info('Resolve the mask conflict after mask propagate...')
   432 23496.4492 MiB 226.4922 MiB           1               fix_mask_conflict(self.masks, self.graph_module, self.dummy_input)
   433                                         
   434 23496.6914 MiB   0.2422 MiB           1               self.graph_module.load_state_dict(torch.load(ori_state_dict_file.name))
   435 23496.6914 MiB   0.0000 MiB           1               self.graph_module.train(training)
   436                                                 finally:
   437 23496.6914 MiB   0.0000 MiB           1               import os
   438 23496.6914 MiB   0.0000 MiB           1               os.unlink(ori_state_dict_file.name)
   439                                         
   440 23497.0625 MiB   0.3711 MiB           1           self.replace_compressed_modules()
   441 23497.0625 MiB   0.0000 MiB           1           self.logger.info("Speedup done.")
   442                                         
   443                                                 # for node, node_info in self.node_infos.items():
   444                                                     # Pointer.delete_cytpes(node_info.output_masks)
   445                                         
   446                                                     # m_c = len(gc.get_referrers(node_info.output_masks))
   447                                                     # for i in range(len(gc.get_referrers(node_info.output_masks))):
   448                                                     #     del gc.get_referrers(node_info.output_masks)[i]
   449                                                     # for i in range(len(gc.get_referrers(node_info.output_grad))):
   450                                                     #     del gc.get_referrers(node_info.output_grad)[i]
   451                                                     # for i in range(len(gc.get_referrers(node_info.output_origin))):
   452                                                     #     del gc.get_referrers(node_info.output_origin)[i]
   453                                                     # for i in range(len(gc.get_referrers(node_info.output_randomize))):
   454                                                     #     del gc.get_referrers(node_info.output_randomize)[i]
   455                                                     # m_c = sys.getrefcount(node_info.output_masks)
   456                                                     # g_c = sys.getrefcount(node_info.output_grad)
   457                                                     # i_c = sys.getrefcount(node_info._output_inplace)
   458                                                     # del node_info
   459                                                     # sys.getrefcount(node_info)
   460                                         
   461                                                     # for  info in node_info:
   462                                                     #     del info
   463 23497.0625 MiB   0.0000 MiB           1           gc.collect()
   464                                         
   465                                         
   466                                         
   467                                         
   468                                         
   469 15778.5234 MiB -7718.5391 MiB           1           del self.node_infos
   470 15778.5234 MiB   0.0000 MiB           1           del self.logger
   471 15778.5234 MiB   0.0000 MiB           1           del self.masks
   472 15778.5234 MiB   0.0000 MiB           1           del self.graph_module
   473                                                 # del self.bound_model
   474 15778.5234 MiB   0.0000 MiB           1           del self.dummy_input
   475 15778.5234 MiB   0.0000 MiB           1           del self.mask_updaters
   476 15778.5234 MiB   0.0000 MiB           1           del self.replacers
   477 15778.5234 MiB   0.0000 MiB           1           del self.module
   478 15778.5234 MiB   0.0000 MiB           1           del self.submodules
   479 15778.5234 MiB   0.0000 MiB           1           del self.user_to_last_uses
   480 15778.5234 MiB   0.0000 MiB           1           gc.collect()
   481 15778.5234 MiB   0.0000 MiB           1           torch.cuda.empty_cache()
   482                                                 # from pympler import muppy, summary
   483                                                 # all_objects = muppy.get_objects()
   484                                                 # sum1 = summary.summarize(all_objects)  # Prints out a summary of the large objects
   485                                                 # summary.print_(sum1)  # Get references to certain types of objects such as dataframe
   486                                         
   487 15778.5234 MiB   0.0000 MiB           1           return self.bound_model

**Comment by lianxintao** [comment]
/

**Comment by J-shang** [comment]
`speedup.graph_module` also can be del after speedup done.
And you could directly del the whole `speedup` obj after speedup ends.

**Comment by Lijiaoa** [comment]
@lianxintao Could you close this issue if you don't have any questions?
